_posts/2024-09-27-optimizing-assistive-technology.md

```diff
@@ -0,0 +1,200 @@
+user to interact with.
+
+```swift
+VStack {
+    VStack {
+        title
+        description
+    }
+    surveyButton
+}
+.overlay(alignment: .topTrailing) {
+    dismissButton
+}
+.accessibilityElement(children: .combine)
+```
+
+Talking about "with great power comes great responsibility"... the
+`.accessibilityElement(children: .combine)` is _technically_ all we need here,
+but restricts optimizations.
+
+In this example, it does two major things:
+
+- It groups the `Text` children to form its label, like we did with the same
+code on the inner `VStack` previously.
+- It exposes the `Button` children as accessibility actions.
+
+Great..? Well, I'm not entirely sure what users that use VoiceOver would prefer,
+but I was planning to schedule a user interview to find out.
+
+... planning to schedule?
+
+Well, yes. As after making this change, it dawned on me this might make Voice
+Control interaction awkward.
+
+# Voice Control
+
+In VoiceOver, this could be spoken as "We'd love your feedback [..], button,
+actions available". It'd indicate the text, that it is a button (and can be
+interacted with), _and_ that it has (other) actions.
+
+But for Voice Control, we now have a... challenge. How can we interact with
+either button, now we have just one element?
+
+Technically, the whole element is a button. So we could try to activate...
+something that way?
+
+Well, Voice Control will generate an event... in the center of the view. Which
+in our case is just some text. Not the dismiss button, not the survey button.
+
+Well, that's awkward.
+
+Now, _technically_ the user could in this case say "show actions for [element]",
+for which we can at least optimize the `[element]`, which is now that long
+label:
+
+```swift
+VStack {
+    // element details
+}
+.overlay(alignment: .topTrailing) {
+    dismissButton
+}
+.accessibilityElement(children: .combine)
+.accessibilityInputLabels(["Take survey"])
+```
+
+Now the user an at least say "show actions for take survey", and they would then
+be able to either take the survey or dismiss it.
+
+Unfortunately, the only way to visually indicate there are actions for an
+element, is to use "show numbers". And only on iOS[^2] does it indicate a
+"double arrow"... which as far as I know is undocumented, so a user would
+somehow need to be using "show numbers" (and not "show labels") _and_ know what
+this double arrow represents.
+
+But that leaves us with the "ghost" action (which is the most prominent one),
+being "Take survey", which actually doesn't result in doing anything.
+
+I can see a world where "Take survey" would result in actually showing the two
+underlying actions, but that comes with its own array of problems, so perhaps
+it's not that great of an idea in practice.
+
+And so with that, I decided to revert to the version that does combine title and
+description, but otherwise leaves the component alone. Perhaps not the "perfect"
```

Comment:
I wonder if it's helpful to link to the code block you are reverting to, for a reader's easy reference?